### PR TITLE
 Added PlayerUseBowWithoutProjectileEvent #11668 

### PR DESCRIPTION
**Pull request by Chaosdave34**
Added PlayerUseBowWithoutProjectileEvent when a player tries to load a crossbow or draw a bow without having a suitable projectile in their inventory. Makes it possible to set a projectile and enable the player to use the weapon nevertheless.

### CODE REVIEW

File: patches/api/0501-Added-PlayerUseBowWithoutProjectileEvent.patch [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+@@ -0,0 +1,35 @@
++package io.papermc.paper.event.player;
++
++import org.bukkit.entity.Player;
++import org.bukkit.event.HandlerList;
++import org.bukkit.event.player.PlayerEvent;
++import org.bukkit.inventory.ItemStack;
++import org.jetbrains.annotations.NotNull;
++
++public class PlayerUseBowWithoutProjectileEvent extends PlayerEvent {
```
Comment (lynxplay) [inline patches/api/0501-Added-PlayerUseBowWithoutProjectileEvent.patch:26]: `@NullMarked`

File: patches/api/0501-Added-PlayerUseBowWithoutProjectileEvent.patch [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
++
++    private static final HandlerList HANDLER_LIST = new HandlerList();
++
++    private ItemStack projectile;
++
++    public PlayerUseBowWithoutProjectileEvent(@NotNull final Player player) {
```
Comment (lynxplay) [inline patches/api/0501-Added-PlayerUseBowWithoutProjectileEvent.patch]: Remove `@NotNull` as per contributing.md, we use jspecify annotations in new paper types.

File: patches/api/0501-Added-PlayerUseBowWithoutProjectileEvent.patch [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
++        super(player);
++    }
++
++    public ItemStack getProjectile() {
++        return projectile;
```
Comment (lynxplay) [inline patches/api/0501-Added-PlayerUseBowWithoutProjectileEvent.patch]: Requires defensive copy.

File: patches/api/0501-Added-PlayerUseBowWithoutProjectileEvent.patch [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
++    }
++
++    public void setProjectile(ItemStack projectile) {
++        this.projectile = projectile;
```
Comment (lynxplay) [inline patches/api/0501-Added-PlayerUseBowWithoutProjectileEvent.patch]: Requires defensive copy.

**Review by lynxplay — CHANGES_REQUESTED** [review]
Welcome to paper and thank you for your first PR :partying_face: 

Left some basic comments regarding annotations.
The PR is also missing javadocs on the API types.

Beyond that, the event name feels a bit off, the player isn't using the bow there, it is searching for a projectile. We already have some other events there so *maybe* we can look into merging those concepts? Not sure tho.

**Comment by Chaosdave34** [comment]
Maybe PlayerTryUseBowWithoutProjectileEvent would be better?

I already thought about using the PlayerReadyArrowEvent for this.
The simplest way to implement the functionality to let a player fire a bow without having an arrow in their inventory seems to be this for me, because when creating a new ItemStack (even with CraftItemStack#asNmsCopy(), to set the arrow from the CraftItemStack used in the event), it would be diffcult to consume the arrow (removing the arrow from the player inventory) when the arrow ist not manipulated by the event.
(get the arrow ->  create a craftbukkit copy -> call the event with the arrow -> the arrow is not manipulated -> get the arrow from the event -> create a nms copy -> created a new ItemStack although the original one was not changed -> Player#useAmmo() breaks; or am I missing something?)

**Comment by lynxplay** [comment]
No you are not missing anything I don't think.
I don't see a nice way to have this interact with player ready arrow event either no.

Naming wise, well, realistically this should probably not be limited to bows only.
It is a general "PlayerFailedProjectileSearch" or something (shitty name but you get the gist :sweat_smile:) as this is going to be called for crossbows as well.

I'll try to get this PR into a pr party next weekend so I can poke some other maintainers on a better name for this, but I think the functionality this is adding is worth having :+1:

**Pull request closed without merging**